src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,246 @@
+        return false;
+      }
+      logger.debug("Parsed service is as expected");
+      return true;
+    }
+  }
+
+  /**
+   * Tries to find the team project information in the given request or as part of the
+   * stored this.authorizer.getCurrentTeamProjectRoleForCurrentUser(). Returns null
+   * if nothing can be found.
+   */
+  private String extractTeamProjectFromRequestParameters(ServletRequest request) throws Exception {
+    // Get the url
+    HttpServletRequest httpRequest = (HttpServletRequest) request;
+    String url = httpRequest.getRequestURL().toString();
+
+    RoleEntity currentTeamProjectRole = this.authorizer.getCurrentTeamProjectRoleForCurrentUser();
+    String currentTeamProjectName = null; 
+    if (currentTeamProjectRole != null) {
+      currentTeamProjectName = this.authorizer.getCurrentTeamProjectRoleForCurrentUser().getName();
+      if (currentTeamProjectName == null || currentTeamProjectName.trim().isEmpty()) {
+        throw new Exception("The teamproject role was found but name was unexpectedly empty");
+      }
+    }
+    logger.debug("Current teamproject: {}...", currentTeamProjectName);
+    logger.debug("Checking if a teamproject has been specified in the request...");
+
+    // try to find it in the redirectUrl parameter:
+    logger.debug("Looking for redirectUrl in request: {}....", url);
+    String[] redirectUrlParams = getParameterValues(request, "redirectUrl");
+    if (redirectUrlParams != null) {
+      logger.debug("Parameter redirectUrl found. Checking if it contains teamproject....");
+      // teamProject will be in first one in this case...as only parameter:
+      String firstParameter = redirectUrlParams[0].toLowerCase();
+      if (firstParameter.contains("teamproject=")) {
+        String teamProject = firstParameter.split("teamproject=")[1];
+        logger.debug("Found teamproject: {}", teamProject);
+        return teamProject;
+      }
+    }
+
+    // try to find "teamproject" param in url itself (there will be no redirectUrl if user session is still valid):
+    logger.debug("Fallback1: Looking for teamproject in request: {}....", url);
+    String[] teamProjectParams = getParameterValues(request, "teamproject");
+    if (teamProjectParams != null) {
+      logger.debug("Parameter teamproject found. Parsing....");
+      String teamProject = teamProjectParams[0].toLowerCase();
+      logger.debug("Found teamproject: {}", teamProject);
+      return teamProject;
+    }
+
+    logger.debug("Fallback2: Looking for teamproject in Action-Location header of request: {}....", url);
+    String actionLocationUrl = httpRequest.getHeader("Action-Location");
+    if (actionLocationUrl != null && actionLocationUrl.contains("teamproject=")) {
+      String teamProject = actionLocationUrl.split("teamproject=")[1];
+      logger.debug("Found teamproject: {}", teamProject);
+      return teamProject;
+    }
```

Comment:
TODO - cleanup. These are likely not needed by now, and probably just dead code...

Reply:
or maybe not... some fallback scenarios might be needed in cases where the login was already made. The use cases logging in for first time vs opening again after changing team project selection probably need these different implementations.  Maybe not fallback 1 and 2...but probably at least one of them.